### ISSUE TITLE
tree.refresh event runs after initial tree is loaded - quick question

### ISSUE THREAD

**Issue by WriterStat** [mbraak/jqTree#572]
Hi Marco,

I just noticed that the tree.refresh event runs after the json is initially loaded in the tree area.

I'm using dataUrl to load the tree json.

```
dataUrl: loadURl
```
The tree.refresh event isn't listed in the events docs. But I've been using it to know when the tree nodes move as a catch all.

When/on what things does the tree.refresh event trigger/run?

Does the refresh need to run after the tree is already completed and loaded?

What is the best event(s) to use when tracking a node move, a node added, a node name changed, a node deleted, etc...? I had thought that refresh might trigger when the underlying tree data was refreshed.

Thanks!

I'm looking for an event that gets triggered when any of the underlying tree data is changed/refreshed. Is there a best event(s) to watch to catch tree changes/save tree data?

Thanks again,

**Comment by WriterStat** [comment]
I'm currently triggering a save when I add or delete a node, or update a node name and so on in each separate code area etc... and when a node refreshes and so on... but this triggers a save on the very initial load of a tree.

Is there a tree.change event? Thanks again! Great software.

**Comment by mbraak** [comment]
Hello WriterState,

The `tree.refresh` event is fired after the tree is drawn.

Examples when the `tree.refresh` event is fired:

* after the first draw of the tree
* after a node is moved
* after the `updateNode` method is called

I will add documentation about the event.

There is a [tree.move](https://mbraak.github.io/jqTree/#event-tree-move) event, which is called when a node is moved using drag-and-drop.

There is no `tree.change` or `tree.delete` event.

A possible solution is perhaps to trigger the events yourself:

```js
const removeNode = node => {
  $tree.tree('removeNode', node);

  const event = jQuery.Event('tree.delete', { id: node.id });
  $tree.trigger(event);
};
```

And then you could call `removeNode` everywhere you remove a node. Of course, I don't know if this the best solution in your situation.

Greetings,
Marco

**Comment by WriterStat** [comment]
Hi Marco, Thanks!

And for the extra info about the Refresh event.

I was currently triggering an event for each tree function when I add a node, delete a node, update a node name and so on. And using tree.refresh when I move a node. I had added an event trigger to each one.

It just felt a smidge inelegant being every where. So I was thinking about maybe trying a quicker solution based on when the underlying tree js data object changed. This would be valuable as there are some things you want to happen when the tree js data object changes, but you don't really care about the specific change that was made.

Like saving the tree or updating or refreshing other things based on the change.

**Comment by WriterStat** [comment]
I noticed that any external event code you place or code on the tree or inside the tree or on any items/elements placed in the tree have to be placed inside a tree.refresh event function for them to work.

```
$tree.on('tree.refresh', function(e) {
```
Any external items placed in the tree, buttons, inputs, search, text, names, etc...have to have their click, input, and other events coded and placed inside a tree.refresh event for them to work.

Made me scratch my head for a bit when my click events didn't seem to work inside the tree. =) It's how I found the refresh event.
